### PR TITLE
[TRTLLM-14304][feat] Integrate embeddings cache with encoder side-stream

### PR DESCRIPTION
**Pull request by 2ez4bz**
## Dev Engineer Review

- Refactors Gemma4 and Qwen3 multimodal wrappers around shared APIs, enabling encoder-cache integration and preserving multimodal item ordering during cross-iteration prefetch.
- Adds encoder-cache support declarations for Qwen3, Qwen3 MoE, and Qwen3.5 models.
- Improves multimodal token-ID compatibility by preferring `multimodal_token_ids` with legacy `mm_token_ids` fallback.
- Tightens active-token filtering and updates Qwen3 deepstack fusion and embedding behavior.
- Updates Qwen Image Bench configuration to disable encoder caching while preserving runtime metadata.
- Documentation now describes supported models and combined side-stream/cache memory behavior.
- Review areas: validate the substantial Gemma4 inheritance and weight-loading changes, confirm the new `get_language_model_extra_forward_kwargs` API is consistent across wrappers, and verify the single-modality-per-request restriction does not regress supported use cases.
- CI has reported failures for multiple commits and should be investigated before merge.

## QA Engineer Review

- Added or updated coverage for:
  - `test_cross_iter_prefetch_materializes_on_side_stream`
  - `test_prepare_multimodal_indices_uses_mixin_token_ids`
  - `test_prepare_multimodal_indices_uses_legacy_token_ids`
  - `test_encoder_cache_reuses_image_embedding_across_requests`
  - `test_chunked_prefill_reuses_cached_vision_embeddings`
  - `test_mixed_modality_batch_preserves_request_order`
  - `test_chunk_without_multimodal_tokens_is_inactive`
  - `test_single_request_with_multiple_modalities_raises`
  - `test_qwen3vl_deepstack_fusion_reuses_registered_buffer`
- No test-list changes were identified in the provided summary. CI/test-list coverage for these tests should be confirmed; the reported L0 failures require follow-up.
- Verdict: needs follow-up.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/models/modeling_multimodal_mixin.py [hunk 1]
```diff
@@ -52,6 +52,26 @@
 _MM_ENCODER_CACHE_LOG_NAME = "mm_encoder_cache"
 
 
+def _build_request_multimodal_input(
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/models/modeling_multimodal_mixin.py:228]: This was moved from `tensorrt_llm/_torch/pyexecutor/model_engine.py` to avoid circular dependencies.

File: tests/unittest/_torch/multimodal/test_multimodal_mixin.py [hunk 2]
```diff
@@ -255,6 +259,16 @@ def test_encoder_cache_first_request_writes_per_item_entries():
...
 
 
+def test_encoder_cache_requires_model_opt_in():
+    model = DummyMultimodalModel(make_embedding(hidden_size=4), torch.tensor([7]))
+    model.model_config = ModelConfig(
+        multimodal_config=MultimodalConfig(encoder_cache_max_bytes=4096)
+    )
+
+    assert not model._encoder_cache_active
+    assert model._get_multimodal_encoder_cache() is None
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tests/unittest/_torch/multimodal/test_multimodal_mixin.py]: Done.

File: tests/unittest/_torch/multimodal/test_multimodal_mixin.py [hunk 3]
```diff
@@ -255,6 +259,16 @@ def test_encoder_cache_first_request_writes_per_item_entries():
     assert len(model._multimodal_encoder_cache) == 2
 
 
+def test_encoder_cache_requires_model_opt_in():
+    model = DummyMultimodalModel(make_embedding(hidden_size=4), torch.tensor([7]))
+    model.model_config = ModelConfig(
+        multimodal_config=MultimodalConfig(encoder_cache_max_bytes=4096)
+    )
+
+    assert not model._encoder_cache_active
```
Comment (cascade812) [inline tests/unittest/_torch/multimodal/test_multimodal_mixin.py]: nit: does _encoder_cache_active exist?

**Review by cascade812 — APPROVED** [review]

File: tests/unittest/_torch/multimodal/test_multimodal_mixin.py (hunk 3 above)
Comment (2ez4bz) [reply tests/unittest/_torch/multimodal/test_multimodal_mixin.py]: Woops, remnant of a prior implementation. Fixed, thanks.

**Review by arysef — APPROVED** [review]

File: tests/unittest/_torch/test_tensor_lru_cache.py [hunk 4]
```diff
@@ -180,3 +180,82 @@ def write_and_read(index: int) -> None:
         hit = cache.get(index)
         if hit is not None:
             assert hit.numel() * hit.element_size() == 8
+
+
+def test_stream_aware_mode_leaves_cpu_cache_behavior_unchanged() -> None:
```
Comment (brnguyen2) [inline tests/unittest/_torch/test_tensor_lru_cache.py:185]: Noticed `unittest/_torch/test_tensor_lru_cache.py` isn't registered in any
test-db list: `grep -rn "tensor_lru" tests/integration/test_lists/` comes back
empty, and there's no bare `unittest/_torch` directory glob (top-level files in
that directory are listed individually, e.g. `unittest/_torch/test_model_config.py`
in `l0_a10.yml`). So neither the pre-existing tests nor the new stream-safety
ones run in CI.

That gap predates this PR, but the new CUDA tests are exactly the ones guarding
the cross-stream race, so it seems worth actually running them pre-merge. The
whole file takes well under a second (the `_sleep` calls are ~10ms each). Add it
to the A10 pre-merge unittest block in `l0_a10.yml`?

File: tests/unittest/_torch/test_tensor_lru_cache.py [hunk 5]
```diff
@@ -180,3 +180,82 @@ def write_and_read(index: int) -> None:
...
+        assert popped is not None
+        observed = popped.clone()
+
+    consumer_stream.synchronize()
+    assert len(cache) == 0
+    torch.testing.assert_close(observed, torch.full_like(observed, 5))
+
+
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="requires CUDA")
+def test_cuda_stream_aware_cache_preserves_evicted_consumer_storage() -> None:
```
Comment (brnguyen2) [inline tests/unittest/_torch/test_tensor_lru_cache.py:241]: This test exercises same-key replacement (`put()` on an existing key pops the
old entry directly) rather than capacity-driven eviction through
`_evict_until_within_limit()`, which is the path the comment in
`tensor_lru_cache.py` calls out ("an evicted tensor may still have outstanding
consumers on another stream"). The reference-drop mechanics are the same, so no
bug suspected, but putting a different key here would force actual LRU eviction,
cover that branch, and make the test match its name.

`clear()` in stream-aware mode is similarly uncovered, if you want to fold that
into the same change.

**Review by brnguyen2 — APPROVED** [review]

File: tests/unittest/_torch/test_tensor_lru_cache.py (hunk 4 above)
Comment (2ez4bz) [reply tests/unittest/_torch/test_tensor_lru_cache.py:185]: Good catch, that's definitely an oversight - fixed.

File: tests/unittest/_torch/test_tensor_lru_cache.py (hunk 5 above)
Comment (2ez4bz) [reply tests/unittest/_torch/test_tensor_lru_cache.py:241]: Done.

File: tests/unittest/_torch/multimodal/test_mm_encoder_cross_iter_prefetch.py [hunk 6]
```diff
@@ -242,6 +308,102 @@ def test_cross_iter_prefetch_materializes_on_side_stream(monkeypatch):
...
+    assert model.encoder_call_count == 2
+    assert model.last_encoder_batch_size == 1
+    torch.testing.assert_close(
+        hit.py_multimodal_data["multimodal_embedding"],
+        torch.ones((4, 8), device="cuda"),
+    )
+    torch.testing.assert_close(
+        miss.py_multimodal_data["multimodal_embedding"],
+        torch.full((4, 8), 2.0, device="cuda"),
+    )
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tests/unittest/_torch/multimodal/test_mm_encoder_cross_iter_prefetch.py:416]: Both requests share the same event, recorded after all candidate work, so synchronizing hit.py_mm_encoder_event also covers miss. No additional synchronization is needed, I believe.

File: tests/unittest/_torch/multimodal/test_mm_encoder_cross_iter_prefetch.py [hunk 7]
```diff
@@ -259,3 +421,112 @@ def test_cross_iter_prefetch_skips_in_flight_and_cached(monkeypatch):
...
+    monkeypatch.setattr(torch.cuda, "synchronize", lambda *a, **k: sync_calls.append("device"))
+    monkeypatch.setattr(
+        torch.cuda.Stream, "synchronize", lambda self, *a, **k: sync_calls.append("stream")
+    )
+
+    n = maybe_prefetch_mm_encoder_for_next_iter(model, [req], max_prefetch_ahead=2)
+
+    assert n == 1
+    assert sync_calls == []
+    assert not req.py_mm_encoder_event.query()
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tests/unittest/_torch/multimodal/test_mm_encoder_cross_iter_prefetch.py]: Tweaked the logic, as even higher cycle counts didn't reliably work. Check again.

[the remark this answers is not in the document]
Comment (2ez4bz) [reply tests/unittest/_torch/multimodal/test_mm_encoder_cross_iter_prefetch.py]: I think you may have been looking at an outdated tree. Please check again.

File: tensorrt_llm/_torch/models/modeling_multimodal_utils.py [hunk 8]
```diff
@@ -178,13 +178,14 @@ def _get_uncached_multimodal_params(
     return params_to_run
 
 
-def _cache_multimodal_embeddings(
```
Comment (yechank-nvidia) [inline tensorrt_llm/_torch/models/modeling_multimodal_utils.py:181]: Is this no-longer doing caching?

File: tensorrt_llm/_torch/models/modeling_multimodal_mixin.py
Comment (yechank-nvidia) [inline tensorrt_llm/_torch/models/modeling_multimodal_mixin.py]: I think we don't need to send the `multimodal_data` for the cache hit.

**Review by StanleySun639 — APPROVED** [review]

File: tests/unittest/_torch/test_tensor_lru_cache.py [hunk 9]
```diff
@@ -180,3 +180,86 @@ def write_and_read(index: int) -> None:
         hit = cache.get(index)
         if hit is not None:
             assert hit.numel() * hit.element_size() == 8
+
```
Comment (StanleySun639) [inline tests/unittest/_torch/test_tensor_lru_cache.py:183]: AI told me that "tests/unittest/_torch/test_tensor_lru_cache.py currently needs an L0 YAML entry." 
Could you double-check and add the entry for it in L0?

Comment (StanleySun639) [reply tests/unittest/_torch/test_tensor_lru_cache.py:183]: Please ignore, I saw you added it.

File: tensorrt_llm/_torch/models/modeling_multimodal_mixin.py
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/models/modeling_multimodal_mixin.py]: For example, we can do something like following:
```
  with _run_on_aux_stream(aux_stream) as encoder_event:
      if encoder_cache_enabled:
          cache = model._get_multimodal_encoder_cache()
          misses = [p for p in params_list
                    if not model._attach_encoder_cache_hit(p, cache)]   # CPU-only
      else:
          misses = params_list

      for p in misses:                       # Only send missed data
          p.to_device("multimodal_data", "cuda", ...)
      for (req, _, _), p in zip(candidates, params_list):
          req.py_multimodal_data = p.multimodal_data

      if misses:
          out = model.encode_multimodal_inputs(misses)
          _store_chunked_prefill_embeddings(misses, [out])
          for p in misses:
              model._write_encoder_cache_entries(p, cache)
```

File: tensorrt_llm/_torch/models/modeling_multimodal_mixin.py [hunk 10]
```diff
@@ -928,8 +980,11 @@ def _dispatch_cross_iter_prefetch(
             # model_engine._prepare_inputs.
             for (req, _, _), p in zip(candidates, params_list):
                 req.py_multimodal_data = p.multimodal_data
-            encoder_output = model.encode_multimodal_inputs(params_list)
-            _cache_multimodal_embeddings(params_list, [encoder_output])
+            if encoder_cache_enabled:
+                model._get_or_encode_multimodal_embeddings(params_list)
```
Comment (yechank-nvidia) [inline tensorrt_llm/_torch/models/modeling_multimodal_mixin.py]: Seems `_get_or_encode_multimodal_embeddings` actually reutrns output embedding, and its return is abandoned while there is a cost of returning embedding(i.e. torch.cat())

Can we modify function of `_get_or_encode_multimodal_embeddings` to not return or skip torch.cat() call when is not needed?

Because, recently I found myself, in TRT-LLM we call torch.cat() call too frequently on Multimodal scenario, and it seems that it calls GPU cuda kernel call everytime when invoked.

File: tensorrt_llm/_torch/models/modeling_multimodal_utils.py (hunk 8 above)
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_multimodal_utils.py:181]: It still caches embeddings within the current request for chunked prefill. I renamed it from `_cache_multimodal_embeddings` to `_store_chunked_prefill_embeddings` to distinguish that request-local storage (for chunked prefill, essentially) from the new, persistent, cross-request encoder cache. The function still writes the embeddings into each param’s `multimodal_data`.

File: tensorrt_llm/_torch/models/modeling_multimodal_mixin.py
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_multimodal_mixin.py]: Nice catch! Cache lookup now happens before H2D transfers.

File: tensorrt_llm/_torch/models/modeling_multimodal_mixin.py (hunk 10 above)
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_multimodal_mixin.py]: I think this should be addressed as part of the change to the other comment you left [here](https://github.com/NVIDIA/TensorRT-LLM/pull/16554#discussion_r3662777628).

Side-stream prefetch now:

* calls `encode_multimodal_inputs()` only for misses
* stores the resulting embeddings directly
* writes misses' embeddings to the cache
* validates requests independently
* does not call `_get_or_encode_multimodal_embeddings()` / perform the final `torch.cat`.

File: tensorrt_llm/_torch/models/modeling_multimodal_mixin.py
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_multimodal_mixin.py]: Created TRTLLM-14726 + referenced as a code comment in `tensorrt_llm/_torch/pyexecutor/model_engine.py`.

**Review by yechank-nvidia — APPROVED** [review]
LGTM
